src/main/java/usi/si/seart/service/GitRepoTopicsService.java

```diff
@@ -0,0 +1,82 @@
+package usi.si.seart.service;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.experimental.FieldDefaults;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.util.ConcurrentReferenceHashMap;
+import usi.si.seart.model.GitRepo;
+import usi.si.seart.model.GitRepoTopic;
+import usi.si.seart.model.Topic;
+import usi.si.seart.repository.GitRepoTopicRepository;
+import usi.si.seart.repository.TopicRepository;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.PersistenceContext;
+import javax.validation.constraints.NotNull;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public interface GitRepoTopicsService {
+
+    List<String> getAllTopicsSortByPopularity();
+    Topic getOrCreateTopic(@NotNull String languageName);
+
+
+    void createOrUpdateGitRepoTopics(@NotNull GitRepo repo, @NotNull List<GitRepoTopic> topics);
+
+    @Service
+    @FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+    @AllArgsConstructor(onConstructor_ = @Autowired)
+    @PersistenceContext
+    class GitRepoTopicsServiceImpl implements GitRepoTopicsService {
+
+        EntityManager entityManager;
+
+        EntityManagerFactory entityManagerFactory;
+
+        ConcurrentReferenceHashMap<String, Object> tagLocks = new ConcurrentReferenceHashMap<>();
+
+        TopicRepository topicRepository;
+        GitRepoTopicRepository gitRepoTopicRepository;
+
+
+        @Override
+        public List<String> getAllTopicsSortByPopularity() {
+            return topicRepository.findAllSortByPopularity().stream()
+                    .map(Topic::getLabel)
+                    .limit(3000)
+                    .collect(Collectors.toList());
```

Comment:
This is not how it's done. All results would be loaded in memory regardless. It's better to define the repository method that excepts a `Pageable` and then define the query limit through it!